linked_list/single_linked_list.c

```diff
@@ -0,0 +1,311 @@
+#include <stdio.h>
+#include <stdlib.h>
+#define ASSERT(x) printf("Assertion: " #x " = %d\n", (x))
+
+typedef int data_t;
+
+typedef struct node {
+  data_t data;
+  struct node *next;
+} node;
+
+#define CHECK_ALLOC(ptr)						\
+  if (!ptr) {								\
+    printf("Error allocating memory for %s in  %s: %d line", #ptr, __func__, __LINE__-1); \
+    exit(2);								\
+  }
+
+node* list_create(data_t data){
+  node* first = (node*)malloc(sizeof(node));
+  CHECK_ALLOC(first);
+  first->data = data;
+  first->next = NULL;
+  return first;
+}
+
+
+void list_free(node** head){
+  node* current = *head;
+  while (current != NULL)
+    {
+      node* tmp = current;
+      current = current->next;
+      free(tmp);
+    }
+  *head = NULL; 
+}
+
+void list_append(node** list, int new_element){
+  node* new_node = (node*)malloc(sizeof(node)); 
+  CHECK_ALLOC(new_node); 
+  new_node->data = new_element;
+  new_node->next = NULL;
+
+  if (*list == NULL){
+    *list = new_node;
+  }
+  else {
+    node* current = *list;
+    while (current->next)
+      current = current->next;
+      current->next = new_node;
+    }
+}
+
+
+int list_pop(node** list){
+  int result;
+  
+  if (!(*list))
+    return -1;
+
+  if (!(*list)->next) {
+    result = (*list)->data;
+    free(*list);
+    *list = NULL;
+    return result;
+  }
+  else {
+    node* prev_node = NULL;
+    while ((*list)->next != NULL) {
+      prev_node = *list;
+      *list = (*list)->next; 
+    } 
+  
+    if (prev_node)
+      prev_node->next = NULL;
+
+    result = (*list)->data; 
+    free(*list);
+    return result;
+  }
```

Comment:
Also add a comment above the function, about "how it fails".

(Ofc in a real implementation it would have to return some kind of error, because currently it works similar to `atoi`).